### DISCUSSION TITLE
Events vs V2X Messages

### DISCUSSION THREAD

**Discussion by adrs221** [eclipse/mosaic#71] [Q&A]
I am having a difficult time understanding the difference between CAM/DENM messages and Events. It looks like Roadside Units and Servers can't directly receive V2X messages, since they don't have the OnMessageReceived() method.

It seems like the processEvent() method is what needs to be used here, for example, for the server to generate DEN messages. How do I go about turning CAM and sensor information into an event that the RSU's and server can process?

I looked through the example applications and am confused by what the EventProcessor does/how to create one.

Thanks in advance.

**Comment by rprotzmann** [comment]
`Events` in the Application Simulator are the basic means for

- information exchange (the event includes time + resource) or
- merely increasing time (the event only includes time)

Events could be created internally – by the same app, which will process the event; or even by other applications. Moreover, events could be created due to `Interactions` – typically subscribed by the Application Simulator and published by other simulators.

`V2XMessages` and more specifically CAMs / DENMs are the information of events and accordingly encapsulated in the event resource and scheduled at their according timestamp – see [Documentation – application events](https://www.eclipse.org/mosaic/docs/develop_applications/event_scheduler/).


For Event handling, the concept provides callback methods. For instance, the `processEvent()` is called for general events. For most important Events such as `VehicleUpdates` or the mentioned reception of `V2XMessages`, dedicated callbacks are implemented.

An RSU application, which would like to be called back `onMessageReceived()` needs to be a `CommunicationApplication` and implement the according interface, i.e.
`public class MyApp extends AbstractApplication<RoadSideUnitOperatingSystem> implements CommunicationApplication { }`

An overview of all application interfaces could be found in the [Documentation - application simulator](https://www.eclipse.org/mosaic/docs/simulators/application_simulator/#application-interfaces)
